Tests/individual multi genome

mschmidt87 opened the pull request:
This PR addresses #133 . It simply executes all tests for both `IndividualSingleGenome` and `IndividualMultiGenome` by wrapping the `Genome` into a list for the multi-genome tests.

To keep the additional code to a minimum, I introduced some helper functions.

Edit:
Additionally, I reorganized the tests to use pytest fixtures which makes the test more modular and extending them more scalable. Essentially, the `to_sympy` etc. tests have three dimensions of parameters:
- `individual_type`
- `params` = the combination of genome, primitives, dna, and target function to test for
- `arg_funcs` = the function arguments that the functions are tested for

Edit 2:
I also added a similar test for `to_numpy` which was lacking before.

jakobj commented on test/test_individual.py:
```diff
@@ -3,21 +3,46 @@
 
 import cgp
-from cgp.individual import IndividualSingleGenome
+from cgp import Genome, IndividualSingleGenome, IndividualMultiGenome
```
why this `from` import? couldn't use those classes with the `cgp.` prefix as well?

mschmidt87 replied:
yes, it just saves characters in the code.

jakobj commented on test/test_individual.py:
```diff
@@ -3,21 +3,46 @@
 
 
-def test_pickle_individual():
+def _create_individual(fitness, genome, ind_type="single"):
```
maybe be extra verbose and always write `individual` instead of `ind`?

mschmidt87 replied:
okay, implemented.

mschmidt87 replied:
9e92d2814a90543959b922660f462987ddb538fc

jakobj commented on test/test_individual.py:
```diff
@@ -3,21 +3,46 @@
+    elif ind_type == "multi":
+        if isinstance(genome, Genome):
+            genome = [genome]
+        return IndividualMultiGenome(fitness, genome)
```
`else` branch should raise `NotImplementedError`

mschmidt87 replied:
9e92d2814a90543959b922660f462987ddb538fc

jakobj commented on test/test_individual.py:
```diff
@@ -3,21 +3,46 @@
+    if ind_type == "single":
+        return IndividualSingleGenome(fitness, genome)
+    elif ind_type == "multi":
+        if isinstance(genome, Genome):
```
why do we need this `if`? shouldn't there be only one way to use this function to create `IndividualMultiGenome` either with genome being a `Genome` or `[Genome]`?

mschmidt87 replied:
I guess, I was being overly cautious or wanted to prepare possible future cases where we could also pass a list of Genome. Removed.

mschmidt87 replied:
44f5d161a2fcdf1c96c6afc278c37b683f7f3675

jakobj commented on test/test_individual.py:
```diff
@@ -3,21 +3,46 @@
+    if ind_type == "single":
+        return value
+    elif ind_type == "multi":
+        return value[0]
```
(see above, missing `else`)

mschmidt87 replied:
9e92d2814a90543959b922660f462987ddb538fc

jakobj commented on test/test_individual.py:
```diff
@@ -3,21 +3,46 @@
+    if ind_type == "single":
+        return individual.genome
+    elif ind_type == "multi":
+        return individual.genome[0]
```
(see above, missing `else`)

mschmidt87 replied:
9e92d2814a90543959b922660f462987ddb538fc

jakobj commented on test/test_individual.py:
```diff
@@ -158,25 +185,26 @@ def test_to_and_from_torch_plus_backprop():
     def f_target(x):
         return math.pi * x
 
     f = individual.to_torch()
```
shouldn't this just use the `_unpack_evaluation`?

mschmidt87 replied:
This is a bit tricky but the problem is that we need the packed `f` for the `update_parameters_from_torch_class` call below.

jakobj replied:
ah, i see. and creating two instances is not possible, as you need the optimizer and the `update_...` function to work on the same one, right? ok, let's leave it like this.

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
+    elif individual_type == "MultiGenome":
+        return value[0]
+    else:
+        raise NotImplementedError
```
```suggestion
        raise NotImplementedError()
```
it think you need to call the exception, don't you?

mschmidt87 replied:
No, `NotImplementedError` is a shortcut for `NotImplementedError()`. We just need to call it if we want to print a message.

jakobj replied:
cool, i learned something! :) but wouldn't it make sense to provide a message? ;)

mschmidt87 replied:
Fair point. Added a message. 27ff50681435a44ab24f87dc6efef799c55d26a3

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
+    elif individual_type == "MultiGenome":
+        return individual.genome[0]
+    else:
+        raise NotImplementedError
```
(see above)

mschmidt87 replied:
see above.

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
+@pytest.mark.parametrize("params", params_list)
+@pytest.mark.parametrize("func_args", func_args_list)
+def test_individual_with_parameter_sympy(individual_type, params, func_args):
     sympy = pytest.importorskip("sympy")  # noqa
```
```suggestion
    pytest.importorskip("sympy")
```

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
 
 
-def test_pickle_individual():
+TestParams = namedtuple("TestParams", ["genome_params", "primitives", "dna", "target_function"])
```
why are you using a `namedtuple` instead of a dict here? is it much cleaner to offset the overhead of introducing it?

jakobj replied:
also, having this nice modularity, shouldn't we use it in the tests, i.e., introduce an additional "TestParams" instance?

mschmidt87 replied:
We could use a dict, of course, but I think this is the perfect use case for a `NamedTuple`. By using a `NamedTuple`, the code for unpacking it in the tests, gets shorter, because we can treat it as a tuple.

mschmidt87 replied:
> also, having this nice modularity, shouldn't we use it in the tests, i.e., introduce an additional "TestParams" instance?

I don't understand what you mean here, sorry.

jakobj replied:
well, you define a list of `TestParams` but right now this list only has one entry, e.g., tests only one graph version and target function. wouldn't it make sense to test multiple ones?

mschmidt87 replied:
Yes, but I would postpone that to a different PR. Let's not overload this one.

mschmidt87 replied:
Added #182 for this.

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
+        target_function=lambda x, c: x + c,
+    )
+]
+FuncArgs = namedtuple("FuncArgs", ["x", "c"])
```
(see above)
also, i'd rename this to something more specific like `graph_input_values`.

mschmidt87 replied:
Thanks, renamed. Note that namedtuples fall under classes when it comes to naming conventions, therefore `GraphInputValues`. 

Changed in 6cf310c53036ba0d3cb1057fc0a8d837bac938d8

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
     primitives = (cgp.Add,)
     genome = cgp.Genome(1, 1, 1, 1, 1, primitives)
-    individual = IndividualSingleGenome(None, genome)
+    individual = _create_individual(None, genome, individual_type)
```
maybe you could provide a default value for `fitness` s/t you don't need to provide `None`?

mschmidt87 replied:
why? I think it would be misleading to put a value here, if it is no needed.

jakobj replied:
sorry, my comment wasn't clear enough: i mean in the `_create_individual` function you could provide a default value such that you don't need to provide anything here

mschmidt87 replied:
Yes, good idea, added that in 7772699715c2e633bdace8fda363eefe9f80dca9

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
-        ID_NON_CODING_GENE,
-    ]
-    individual = IndividualSingleGenome(None, genome)
+    x, c = func_args.x, func_args.c
```
i think an `assert` got lost that made sure than before setting the parameter in the next line, it's value is equal to the initial value, `1.0`. this also applies for more tests below.

mschmidt87 replied:
Sorry, I don't see that `assert` in the old code (left side of here). As far as I can see, the asserts only tested for the function evaluation, e.g. in line 47 and 55 of the old version.

jakobj replied:
yes, you are right. i was misinterpreting the two consecutive asserts in the old cold: in both cases the parameter is assigned and then the output is checked. since you implemented this via `parametrized` i got a bit confused. since defaults are checked in `test_nodes`, never mind, everything's fine here :+1:

jakobj commented on test/test_individual.py:
```diff
@@ -1,139 +1,165 @@
-    y = f.subs("x_0", x[0]).evalf()
 
-    assert y == pytest.approx(x[0] + c)
+@pytest.mark.parametrize("individual_type", ["SingleGenome", "MultiGenome"])
```
good catch that this was missing! :ok_hand:

mschmidt87 commented:
Thank you for your comments @jakobj , I addressed all of them, please take another look.

mschmidt87 commented:
Okay, I addressed the remaining comments.

I also added a change to the test utils to make this test more robust (currently failing in CI here), ff097b3397211b93396487f03650c1b22667e891

Also note: the coverage display is just wrong. I don't know what coveralls is doing here, but the coverage increased by this PR. We might just get rid of coveralls altogether, to be honest.

jakobj commented on test/test_utils.py:
```diff
@@ -54,7 +54,7 @@ def evaluate_objective_on_list(x):
     t0 = time.time()
     evaluate_objective_on_list(x)
-    assert (time.time() - t0) < (sleep_time / 5.0)
+    assert (time.time() - t0) < pytest.approx(sleep_time / 5.0, rel=0.1)
```
hmm, travis doesn't seem to like this, and i'd also choose a different route: we should just increase the `sleep_time` a bit, maybe to 2s. this test fails as travis only gives us a single core and running multiple threads in it seems to slow everything down to the point where the effect of caching is not visible any more.

mschmidt87 replied:
Okay, I reverted the change. Miraculously, the CI seems to pass this test now. I created this ticket to refactor the test: #185

mschmidt87 commented:
I reverted the change and squashed all commits. Please approve @jakobj

jakobj commented:
good timing :) we can merge once travis is happy :+1:
